src/base/static/components/templates/main-map.js

```diff
@@ -0,0 +1,368 @@
+      this.props.mapViewport.zoom !== prevProps.mapViewport.zoom
+    ) {
+      this.props.setSlippyRoute();
+    }
+  }
+
+  beginFeatureQuery = evt => {
+    // Relying on react-map-gl's built-in onClick handler produces a noticeable
+    // lag when clicking around Places on the map. It's not clear why, but we
+    // get better performance by querying rendered features as soon as the
+    // onMouseDown or onTouchStart events fire, and using the onMouseUp and  
+    // onTouchEnd handler to test if the most recent queried feature is one we 
+    // shoud route to (i.e. is a Place).
+    //
+    // Note that if no features are found in the query, an empty array is
+    // returned.
+    this.queriedFeatures = this.map.queryRenderedFeatures(evt.point);
+  };
+
+  endFeatureQuery = () => {
+    if (
+      !this.props.isMapDragging &&
+      this.queriedFeatures.length &&
+      this.queriedFeatures[0].properties &&
+      this.queriedFeatures[0].properties._clientSlug
+    ) {
+      // If the topmost clicked-on feature has a _clientSlug property, there's
+      // a good bet we've clicked on a Place. Assume we have and route to the
+      // Place's detail view.
+      this.props.router.navigate(
+        `/${this.queriedFeatures[0].properties._clientSlug}/${this.queriedFeatures[0].id}`,
+        { trigger: true },
+      );
+    }
+  };
+
+  onInteractionStateChange = evt => {
+    if (evt.isDragging && !this.props.isMapDragging) {
+      this.props.updateMapDragging(true);
+    } else if (!evt.isDragging && this.props.isMapDragging) {
+      this.props.updateMapDragging(false);
+    }
+  };
+
+  onMapLoad = () => {
+    this.setState({
+      isMapLoaded: true,
+    });
+  };
+
+  render() {
+    return (
+      <MapGL
+        ref={this.mapRef}
+        mapOptions={this.props.mapConfig.options.mapOptions}
+        width={this.props.mapViewport.width}
+        height={this.props.mapViewport.height}
+        latitude={this.props.mapViewport.latitude}
+        longitude={this.props.mapViewport.longitude}
+        pitch={this.props.mapViewport.pitch}
+        bearing={this.props.mapViewport.bearing}
+        zoom={this.props.mapViewport.zoom}
+        transitionDuration={this.props.mapViewport.transitionDuration}
+        transitionInterpolator={this.props.mapViewport.transitionInterpolator}
+        transitionEasing={this.props.mapViewport.transitionEasing}
+        mapboxApiAccessToken={MAP_PROVIDER_TOKEN}
+        minZoom={this.props.mapViewport.minZoom}
+        maxZoom={this.props.mapViewport.maxZoom}
+        onMouseUp={this.onMouseUp}
+        onMouseDown={this.onMouseDown}
+        onTouchEnd={this.onMouseUp}
+        onTouchStart={this.onMouseDown}
+        onViewportChange={viewport => {
+          this.props.updateMapViewportFromReactMapGL(
+            viewport,
+            this.isMapTransitioning
+              ? false
+              : this.props.mapConfig.options.scrollZoomAroundCenter,
+          );
+        }}
```

Comment:
Related to my earlier comment about removing `updateMapViewportFromReactMapGL`, I think we can simplify things by re-using our `updateMapViewport` action creator.


```js
 // stripping out width and height here becuase the map may resize during a
 // transition (ie: the Place detail view is opened)
 // NOTE: After a re-design, we wouldn't have to do this because our components
 // would "float" on top of the map, letting the map shine through eg: Google Maps
 const { width, height, ...rest } = viewport;
 this.props.updateMapViewport(
            rest,
            this.isMapTransitioning
              ? false
              : this.props.mapConfig.options.scrollZoomAroundCenter,
          );
        }}
```

and we'd update the `updateMapViewport` action creator from this:

```js
export function updateMapViewport(viewport) {
//etc
payload: { viewport, scrollZoomAroundCenter: false },
//etc
```

to this:

```js
export function updateMapViewport(viewport, scrollZoomAroundCenter=false) {
// etc
payload: { viewport, scrollZoomAroundCenter },
// etc
```

It seems simpler to me, but it's kind of a nit. I'll leave it up to you whether you think it's worth the change.

Reply:
Ah, I see-- yes, I think this is simpler, and there's less code duplication. 👍